File: data/derived_cot/rq3_prompting/results/translation/qwen/cpp_to_py/code_output/Snake.py
import random
import time

class Snake:
    def __init__(self, screenWidth, screenHeight, blockSize, foodPosition):
        self.length = 1
        self.SCREEN_WIDTH = screenWidth
        self.SCREEN_HEIGHT = screenHeight
        self.BLOCK_SIZE = blockSize
        self.positions = [(screenWidth // 2, screenHeight // 2)]
        self.score = 0
        self.food_position = foodPosition

    def move(self, direction):
        cur = self.positions[0]
        x, y = direction
        new_pos = ((cur[0] + x * self.BLOCK_SIZE) % self.SCREEN_WIDTH,
                   (cur[1] + y * self.BLOCK_SIZE) % self.SCREEN_HEIGHT)
        
        if new_pos == self.food_position:
            self.eat_food()
        else:
            if len(self.positions) > 2 and new_pos in self.positions[2:]:
                self.reset()
            else:
                self.positions.insert(0, new_pos)
                if len(self.positions) > self.length:
                    self.positions.pop()

    def random_food_position(self):
        random.seed(int(time.time()))
        valid_position = False
        while not valid_position:
            x = (random.randint(0, (self.SCREEN_WIDTH // self.BLOCK_SIZE) - 1) * self.BLOCK_SIZE
            y = (random.randint(0, (self.SCREEN_HEIGHT // self.BLOCK_SIZE) - 1) * self.BLOCK_SIZE
            self.food_position = (x, y)
            valid_position = True
            for pos in self.positions:
                if pos == self.food_position:
                    valid_position = False
                    break

    def reset(self):
        self.length = 1
        self.positions = [(self.SCREEN_WIDTH // 2, self.SCREEN_HEIGHT // 2)]
        self.score = 0
        self.random_food_position()

    def eat_food(self):
        self.length += 1
        self.score += 100
        self.random_food_position()

    def get_length(self):
        return self.length

    def get_positions(self):
        return self.positions.copy()

    def get_score(self):
        return self.score

    def get_food_position(self):
        return self.food_position

    def get_SCREEN_WIDTH(self):
        return self.SCREEN_WIDTH

    def get_SCREEN_HEIGHT(self):
        return self.SCREEN_HEIGHT

    def get_BLOCK_SIZE(self):
        return self.BLOCK_SIZE

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/py_to_cpp/code_output/Snake.cpp
#include <vector>
#include <utility>
#include <random>
#include <algorithm>

class Snake {
public:
    int length;
    int SCREEN_WIDTH;
    int SCREEN_HEIGHT;
    int BLOCK_SIZE;
    std::vector<std::pair<int, int>> positions;
    int score;
    std::pair<int, int> food_position;

private:
    std::mt19937 rng;

    static int mod(int a, int b) {
        return ((a % b) + b) % b;
    }

public:
    Snake(int SCREEN_WIDTH, int SCREEN_HEIGHT, int BLOCK_SIZE, std::pair<int, int> food_position)
        : length(1),
          SCREEN_WIDTH(SCREEN_WIDTH),
          SCREEN_HEIGHT(SCREEN_HEIGHT),
          BLOCK_SIZE(BLOCK_SIZE),
          positions({{SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2}}),
          score(0),
          food_position(food_position),
          rng(std::random_device{}()) {}

    void move(std::pair<int, int> direction) {
        auto cur = positions[0];
        int x = direction.first;
        int y = direction.second;

        auto new_pos = std::make_pair(
            mod(cur.first + x * BLOCK_SIZE, SCREEN_WIDTH),
            mod(cur.second + y * BLOCK_SIZE, SCREEN_HEIGHT)
        );

        if (new_pos == food_position) {
            eat_food();
        }

        if (positions.size() > 2) {
            bool found = false;
            for (size_t i = 2; i < positions.size(); ++i) {
                if (positions[i] == new_pos) {
                    found = true;
                    break;
                }
            }
            if (found) {
                reset();
                return;
            }
        }

        positions.insert(positions.begin(), new_pos);
        if (static_cast<int>(positions.size()) > length) {
            positions.pop_back();
        }
    }

    void random_food_position() {
        std::uniform_int_distribution<int> dist_x(0, SCREEN_WIDTH / BLOCK_SIZE - 1);
        std::uniform_int_distribution<int> dist_y(0, SCREEN_HEIGHT / BLOCK_SIZE - 1);
        while (std::find(positions.begin(), positions.end(), food_position) != positions.end()) {
            food_position = std::make_pair(dist_x(rng) * BLOCK_SIZE, dist_y(rng) * BLOCK_SIZE);
        }
    }

    void reset() {
        length = 1;
        positions = {{SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2}};
        score = 0;
        random_food_position();
    }

    void eat_food() {
        length += 1;
        score += 100;
        random_food_position();
    }
};